#Lab 3

**1. Normalization and Standardization**

Q1: Normalize the dataset using Min-Max normalization.

In [ ]:
from sklearn.preprocessing import MinMaxScaler
import pandas as pd

df = pd.read_csv('Air_county.csv')

scaler = MinMaxScaler()
normalized_df = pd.DataFrame(scaler.fit_transform(df.select_dtypes(include=['float64', 'int64'])), columns=df.select_dtypes(include=['float64', 'int64']).columns)

normalized_df.head()

Year  Days with AQI  Good Days  Moderate Days  \
0   0.0       0.769014   0.619178       0.177778   
1   0.0       0.309859   0.271233       0.069841   
2   0.0       0.766197   0.608219       0.174603   
3   0.0       0.985915   0.772603       0.234921   
4   0.0       0.659155   0.580822       0.104762   

   Unhealthy for Sensitive Groups Days  Unhealthy Days  Very Unhealthy Days  \
0                             0.014925             0.0                  0.0   
1                             0.000000             0.0                  0.0   
2                             0.044776             0.0                  0.0   
3                             0.037313             0.0                  0.0   
4                             0.000000             0.0                  0.0   

   Hazardous Days   Max AQI  90th Percentile AQI  Median AQI  Days CO  \
0             0.0  0.023655             0.283505    0.222222      0.0   
1             0.0  0.013257             0.257732    0.204678      0.0   
2             0.0  0.029893             0.288660    0.233918      0.0   
3             0.0  0.024435             0.298969    0.233918      0.0   
4             0.0  0.020535             0.247423    0.233918      0.0   

   Days NO2  Days Ozone  Days SO2  Days PM2.5  Days PM10  
0       0.0    0.573770       0.0    0.202186        0.0  
1       0.0    0.000000       0.0    0.330601        0.0  
2       0.0    0.571038       0.0    0.202186        0.0  
3       0.0    0.874317       0.0    0.112022        0.0  
4       0.0    0.669399       0.0    0.000000        0.0

Q2: Standardize the dataset using Z-score standardization.

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
standardized_df = pd.DataFrame(scaler.fit_transform(df.select_dtypes(include=['float64', 'int64'])), columns=df.select_dtypes(include=['float64', 'int64']).columns)

standardized_df.head()

Year  Days with AQI  Good Days  Moderate Days  \
0   0.0      -0.278880   0.014866      -0.339481   
1   0.0      -2.097518  -1.550150      -0.979992   
2   0.0      -0.290037  -0.034426      -0.358319   
3   0.0       0.580231   0.704952      -0.000387   
4   0.0      -0.714014  -0.157656      -0.772768   

   Unhealthy for Sensitive Groups Days  Unhealthy Days  Very Unhealthy Days  \
0                            -0.499562        -0.19158            -0.180833   
1                            -0.652493        -0.19158            -0.180833   
2                            -0.193699        -0.19158            -0.180833   
3                            -0.270165        -0.19158            -0.180833   
4                            -0.652493        -0.19158            -0.180833   

   Hazardous Days   Max AQI  90th Percentile AQI  Median AQI   Days CO  \
0       -0.083454 -0.171800            -0.301242   -0.022252 -0.077882   
1       -0.083454 -0.396586            -0.529153   -0.255136 -0.077882   
2       -0.083454 -0.036929            -0.255660    0.133004 -0.077882   
3       -0.083454 -0.154941            -0.164496    0.133004 -0.077882   
4       -0.083454 -0.239236            -0.620318    0.133004 -0.077882   

   Days NO2  Days Ozone  Days SO2  Days PM2.5  Days PM10  
0 -0.232549    0.355820 -0.246665   -0.298820   -0.26305  
1 -0.232549   -1.401595 -0.246665    0.133612   -0.26305  
2 -0.232549    0.347451 -0.246665   -0.298820   -0.26305  
3 -0.232549    1.276370 -0.246665   -0.602442   -0.26305  
4 -0.232549    0.648722 -0.246665   -0.979670   -0.26305

Q3: Compare the results of normalization and standardization on the Max AQI column. When
would you prefer one technique over the other?

- NOrmalization is better when want to keep values between 0 and 1
- Standardization is better for data that has outliers or needs to be centered around 0

**2. Encoding Categorical Variables**

Q4: Convert the categorical variable State into a numerical form using one-hot encoding. How
many new columns are created, and what are their names?

In [ ]:
import pandas as pd

df = pd.read_csv('Air_county.csv')

df_one_hot = pd.get_dummies(df, columns=['State'])

new_columns = df_one_hot.columns.difference(df.columns)
print(f"Number of new columns created: {len(new_columns)}")
print(f"Names of new columns: {list(new_columns)}")

Number of new columns created: 54
Names of new columns: ['State_Alabama', 'State_Alaska', 'State_Arizona', 'State_Arkansas', 'State_California', 'State_Colorado', 'State_Connecticut', 'State_Country Of Mexico', 'State_Delaware', 'State_District Of Columbia', 'State_Florida', 'State_Georgia', 'State_Hawaii', 'State_Idaho', 'State_Illinois', 'State_Indiana', 'State_Iowa', 'State_Kansas', 'State_Kentucky', 'State_Louisiana', 'State_Maine', 'State_Maryland', 'State_Massachusetts', 'State_Michigan', 'State_Minnesota', 'State_Mississippi', 'State_Missouri', 'State_Montana', 'State_Nebraska', 'State_Nevada', 'State_New Hampshire', 'State_New Jersey', 'State_New Mexico', 'State_New York', 'State_North Carolina', 'State_North Dakota', 'State_Ohio', 'State_Oklahoma', 'State_Oregon', 'State_Pennsylvania', 'State_Puerto Rico', 'State_Rhode Island', 'State_South Carolina', 'State_South Dakota', 'State_Tennessee', 'State_Texas', 'State_Utah', 'State_Vermont', 'State_Virgin Islands', 'State_Virginia'

Q5: Use label encoding for the County column. What are the numerical labels assigned to each
county?

In [ ]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()

df['County_Label'] = label_encoder.fit_transform(df['County'])

county_labels = pd.DataFrame({'County': df['County'], 'County_Label': df['County_Label']}).drop_duplicates()
print(county_labels)

          County  County_Label
0        Baldwin            43
1           Clay           161
2        Colbert           173
3         DeKalb           208
4         Elmore           248
...          ...           ...
1066    Sublette           710
1067  Sweetwater           720
1068       Teton           730
1069       Uinta           751
1070      Weston           788

[819 rows x 2 columns]


Q6: Discuss the impact of using one-hot encoding vs label encoding on the model performance
when the categorical variables are used in regression analysis.

- One-hot encoding treats categories as independent but increases features, which increases memory usage, while label encoding is efficient but imposes an order, which may be misleading in unordered categories.​

**3. Discretization and Binning**

Q7: Bin the Max AQI column into 4 categories: "Low", "Moderate", "High", and "Very High"
using equal-width binning. How many rows fall into each category?

In [ ]:
bins = 4
labels = ["Low", "Moderate", "High", "Very High"]

df['AQI_Category'] = pd.cut(df['Max AQI'], bins=bins, labels=labels)

category_counts = df['AQI_Category'].value_counts()
print(category_counts)

AQI_Category
Low          1067
Moderate        2
Very High       2
High            0
Name: count, dtype: int64


**4. Handling Skewed Data**

Q8: Check the skewness of the Max AQI and Median AQI columns. Which of these columns is
highly skewed?

In [ ]:
max_aqi_skew = df['Max AQI'].skew()
median_aqi_skew = df['Median AQI'].skew()

print(f"Skewness of Max AQI: {max_aqi_skew}")
print(f"Skewness of Median AQI: {median_aqi_skew}")

if abs(max_aqi_skew) > abs(median_aqi_skew):
    print("Max AQI more skewed.")
else:
    print("Median AQI more skewed.")

Skewness of Max AQI: 17.546818819499688
Skewness of Median AQI: 0.6783504567403694
Max AQI more skewed.


Q9: Apply a log transformation to the Max AQI column to handle its skewness. How does this
affect the skewness?

In [ ]:
import numpy as np

df['Log_Max_AQI'] = np.log(df['Max AQI'] + 1)

original_skewness = df['Max AQI'].skew()
log_skewness = df['Log_Max_AQI'].skew()

print(f"Skewness of Max AQI (Original): {original_skewness}")
print(f"Skewness of Max AQI (Log Transformed): {log_skewness}")

Skewness of Max AQI (Original): 17.546818819499688
Skewness of Max AQI (Log Transformed): 0.5141097183673211


Q10: After applying the log transformation, what changes do you observe in the distribution of
the Max AQI data?

- After the log transformation, the skewness of Max AQI dropped from 17.55 to 0.51. This makes the data more evenly distributed

**5. Feature Engineering**

Q11: Create a new feature called Bad_Air_Days, which is the sum of Unhealthy for Sensitive
Groups Days, Unhealthy Days, Very Unhealthy Days, and Hazardous Days. How many counties
have more than 10 Bad_Air_Days?

In [ ]:
df['Bad_Air_Days'] = df[['Unhealthy for Sensitive Groups Days', 'Unhealthy Days', 'Very Unhealthy Days', 'Hazardous Days']].sum(axis=1)

count_bad_air_days = df[df['Bad_Air_Days'] > 10].shape[0]

print(f"Number of counties with more than 10 Bad_Air_Days: {count_bad_air_days}")

Number of counties with more than 10 Bad_Air_Days: 348


**6. Multicollinearity**

Q12: Calculate the correlation matrix for all numeric variables. Which variables are highly
correlated with each other?

In [ ]:
df_encoded = pd.get_dummies(df, columns=['State', 'County'])

correlation_matrix = df_encoded.corr()

print(correlation_matrix)

highly_correlated_pairs = correlation_matrix[(correlation_matrix > 0.8) & (correlation_matrix != 1)]

print("Highly correlated pairs:")
print(highly_correlated_pairs)

                                     Year  Days with AQI  Good Days  \
Year                                  NaN            NaN        NaN   
Days with AQI                         NaN       1.000000   0.731013   
Good Days                             NaN       0.731013   1.000000   
Moderate Days                         NaN       0.506546  -0.188147   
Unhealthy for Sensitive Groups Days   NaN       0.208389  -0.315116   
...                                   ...            ...        ...   
County_Yellowstone                    NaN       0.019444   0.030969   
County_Yolo                           NaN       0.019444   0.009119   
County_York                           NaN      -0.001968  -0.003784   
County_Yukon-Koyukuk                  NaN      -0.066852  -0.042115   
County_Yuma                           NaN       0.019444   0.003092   

                                     Moderate Days  \
Year                                           NaN   
Days with AQI                          

Q13: Based on the correlation matrix, identify pairs of features that might cause multicollinearity.
How could you address this issue?

In [ ]:
threshold = 0.8

correlated_features = []
for column in correlation_matrix.columns:
    for row in correlation_matrix.index:
        if abs(correlation_matrix.loc[row, column]) > threshold and row != column:
            correlated_features.append((row, column))

correlated_features = list(set(tuple(sorted(pair)) for pair in correlated_features))

print("Highly correlated feature pairs:")
for pair in correlated_features:
    print(f"{pair[0]} and {pair[1]}: Correlation = {correlation_matrix.loc[pair[0], pair[1]]}")

Highly correlated feature pairs:
County_Hawaii and Unhealthy Days: Correlation = 0.8540979602610819
County_District of Columbia and State_District Of Columbia: Correlation = 1.0
90th Percentile AQI and Median AQI: Correlation = 0.8155917486744746


- We can address this by removing one of the correlated features like Median AQI or 90th Percentile AQI
- We can also use regularization methods like Ridge or Lasso regression to handle multicollinearity without dropping features

Q14: Remove one of the highly correlated features from the dataset. Recompute the correlation
matrix and compare it with the original one. What changes do you observe?

In [ ]:
df_dropped = df_encoded.drop(columns=['Median AQI'])

new_corr_matrix = df_dropped.corr()

print("\nChanges in the correlation matrix after removing 'Median AQI':")
difference = correlation_matrix.subtract(new_corr_matrix, fill_value=0)

difference = difference[difference != 0].dropna(how='all').dropna(axis=1, how='all')

print(difference)


Changes in the correlation matrix after removing 'Median AQI':
                                     90th Percentile AQI  County_Abbeville  \
90th Percentile AQI                                  NaN               NaN   
County_Abbeville                                     NaN               NaN   
County_Ada                                           NaN               NaN   
County_Adair                                         NaN               NaN   
County_Adams                                         NaN               NaN   
...                                                  ...               ...   
State_Wisconsin                                      NaN               NaN   
State_Wyoming                                        NaN               NaN   
Unhealthy Days                                       NaN               NaN   
Unhealthy for Sensitive Groups Days                  NaN               NaN   
Very Unhealthy Days                                  NaN               NaN   
